### PR TITLE
Implement allowance ranges

### LINKED ISSUE #603: Implement MARECO ranges
When applying a standard allowance on a train run, it will automatically go from the beginning of the path to the end, with the value the user asked for.

Depending on the type of path and/or the rolling stock, the value of the margin might differ on some ranges (for example 5% for 10km, then 4.5mn/100km for 200km, then again 5%, covering the rest of the path).
All combined, theses ranges with different margins cover the entire path.

The user must be able to ask for these specified ranges and the core must compute a train run, allocating the proper allowance on the proper ranges. It includes :

- [x] Design a way to properly transition between two different allowances (in terms of speed)
- [x] Implement different ranges rather than just an AllowanceValue
- [x] Modify the allowance added time that will be the target of the binary search

### PR DESCRIPTION
**Pull request by axrolld**

### CODE REVIEW

File: core/config/checkstyle/checkstyle.xml [hunk 1]
```diff
@@ -300,7 +300,6 @@
                       value="CLASS_DEF, INTERFACE_DEF, ENUM_DEF, METHOD_DEF, CTOR_DEF, VARIABLE_DEF"/>
         </module>
         <module name="JavadocMethod">
-            <property name="accessModifiers" value="public"/>
```
Comment (eckter) [inline core/config/checkstyle/checkstyle.xml]: This isn't needed anymore

File: core/src/main/java/fr/sncf/osrd/envelope_sim/allowances/HardenedMarecoAllowance.java [hunk 2]
```diff
@@ -104,46 +115,60 @@ private Envelope computeMarecoCore(Envelope coreBase, double v1) {
...
+        EnvelopeDeceleration.decelerate(context, envelopeSection.getBeginPos(), beginSpeed, constrainedBuilder, 1);
         return partBuilder.build();
     }
 
-    private EnvelopePart computeSpeedup(Envelope base, double v1) {
-        var endSpeed = base.getEndSpeed();
+    private EnvelopePart computeSpeedup(Envelope envelopeSection, double v1, double targetEndSpeed) {
+        double endSpeed;
+        // if no begin speed has been imposed by the next range
+        if (targetEndSpeed == 0) {
```
Comment (eckter) [inline core/src/main/java/fr/sncf/osrd/envelope_sim/allowances/HardenedMarecoAllowance.java]: 0 means unspecified? I guess it makes sense if we consider that it can't be an imposed speed, but I'd prefer NaN if that works.

File: core/src/main/java/fr/sncf/osrd/envelope_sim/allowances/HardenedMarecoAllowance.java [hunk 3]
```diff
@@ -104,46 +115,60 @@ private Envelope computeMarecoCore(Envelope coreBase, double v1) {
...
+            // if the section is not at the end of the allowance region, there will be no speedup phase
+            if (envelopeSection.getEndPos() < endPos)
```
Comment (eckter) [inline core/src/main/java/fr/sncf/osrd/envelope_sim/allowances/HardenedMarecoAllowance.java:141]: You could say in this comment that transitions between ranges are handled somewhere else, I was confused until I read the rest of the PR.

File: core/src/main/java/fr/sncf/osrd/envelope_sim/allowances/HardenedMarecoAllowance.java [hunk 4]
```diff
@@ -64,6 +65,16 @@ private MarecoSpeedLimit() {
         }
     }
 
+    public static final class CapacitySpeedLimit implements EnvelopeAttr {
+        private CapacitySpeedLimit() {
+        }
```
Comment (eckter) [inline core/src/main/java/fr/sncf/osrd/envelope_sim/allowances/HardenedMarecoAllowance.java]: You don't have to write empty constructors

File: core/src/main/java/fr/sncf/osrd/railjson/schema/schedule/RJSAllowance.java [hunk 5]
```diff
@@ -18,14 +18,16 @@
         public double endPosition = Double.NaN;
         @Json(name = "capacity_speed_limit")
         public double capacitySpeedLimit = -1;
+        @Json(name = "value")
```
Comment (eckter) [inline core/src/main/java/fr/sncf/osrd/railjson/schema/schedule/RJSAllowance.java]:
```suggestion
```

This is only needed when the name is different (generally to go from camelCase to snake_case)

File: core/src/main/java/fr/sncf/osrd/railjson/schema/schedule/RJSAllowance.java [hunk 6]
```diff
@@ -18,14 +18,16 @@
...
         public RJSAllowanceValue value = null;
     }
 
     public static final class Mareco extends RJSAllowance {
         @Json(name = "default_value")
         public RJSAllowanceValue defaultValue = null;
 
-        public RJSRangeAllowance[] ranges = null;
+        @Json(name = "ranges")
```
Comment (eckter) [inline core/src/main/java/fr/sncf/osrd/railjson/schema/schedule/RJSAllowance.java]:
```suggestion
```

File: core/src/main/java/fr/sncf/osrd/railjson/schema/schedule/RJSAllowanceRange.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
+package fr.sncf.osrd.railjson.schema.schedule;
+
+import com.squareup.moshi.Json;
+
+public final class RJSAllowanceRange {
+    @Json(name = "begin_position")
+    public double beginPos;
+    @Json(name = "end_position")
+    public double endPos;
+    @Json(name = "value")
```
Comment (eckter) [inline core/src/main/java/fr/sncf/osrd/railjson/schema/schedule/RJSAllowanceRange.java]:
```suggestion
```

File: core/src/main/java/fr/sncf/osrd/train/StandaloneTrainSchedule.java [hunk 8]
```diff
@@ -13,14 +13,14 @@
 
     public final ArrayList<TrainStop> stops;
 
-    public final ArrayList<? extends Allowance> allowances;
+    public final ArrayList<HardenedMarecoAllowance> allowances;
```
Comment (eckter) [inline core/src/main/java/fr/sncf/osrd/train/StandaloneTrainSchedule.java]: What is the reason for this change? Do we specifically want Mareco here and no other range? (if we eventually have more)

**Review by eckter — CHANGES_REQUESTED** [review]

File: core/src/main/java/fr/sncf/osrd/train/StandaloneTrainSchedule.java (hunk 8 above)
Comment (axrolld) [reply core/src/main/java/fr/sncf/osrd/train/StandaloneTrainSchedule.java]: Nope i guess i just did something wrong here
